File: UnityAdsTests/WebRequestQueueTests.m
#import <XCTest/XCTest.h>
#import "UnityAdsTests-Bridging-Header.h"

@interface WebRequestQueueTests : XCTestCase
@end

@implementation WebRequestQueueTests

- (void)setUp {
    [super setUp];
    [USRVWebRequestQueue start];
}

- (void)tearDown {
    [super tearDown];
}

- (void)testBasicGetRequest {
    __block NSString *targetUrl = [TestUtilities getTestServerAddress];
    XCTestExpectation *expectation = [self expectationWithDescription:@"expectation"];
    
    [USRVWebRequestQueue requestUrl:targetUrl type:@"GET" headers:NULL body:NULL completeBlock:^(NSString *url, NSError *error, NSString *response, long responseCode, NSDictionary<NSString *,NSString *> *headers) {
        XCTAssertEqual(targetUrl, url, "Returned url and requested url should be the same");
        [expectation fulfill];
    } connectTimeout:30000];
    
    __block BOOL success = true;
    [self waitForExpectationsWithTimeout:30 handler:^(NSError * _Nullable error) {
        if (error) {
            success = false;
            XCTAssertTrue(success, "Did not complete");
        }
    }];
    
    
}

- (void)testMultipleGetRequests {
    __block NSString *targetUrl = [TestUtilities getTestServerAddress];
    XCTestExpectation *expectation = [self expectationWithDescription:@"expectation"];
    NSMutableArray *completions = [[NSMutableArray alloc] init];

    [USRVWebRequestQueue requestUrl:targetUrl type:@"GET" headers:NULL body:NULL completeBlock:^(NSString *url, NSError *error, NSString *response, long responseCode, NSDictionary<NSString *,NSString *> *headers) {
        [completions addObject:[NSNumber numberWithInt:1]];
        XCTAssertEqual(targetUrl, url, "Returned url and requested url should be the same");
    } connectTimeout:30000];
    
    [USRVWebRequestQueue requestUrl:targetUrl type:@"GET" headers:NULL body:NULL completeBlock:^(NSString *url, NSError *error, NSString *response, long responseCode, NSDictionary<NSString *,NSString *> *headers) {
        [completions addObject:[NSNumber numberWithInt:2]];
        XCTAssertEqual(targetUrl, url, "Returned url and requested url should be the same");
    } connectTimeout:30000];

    [USRVWebRequestQueue requestUrl:targetUrl type:@"GET" headers:NULL body:NULL completeBlock:^(NSString *url, NSError *error, NSString *response, long responseCode, NSDictionary<NSString *,NSString *> *headers) {
        [completions addObject:[NSNumber numberWithInt:3]];
        XCTAssertEqual(targetUrl, url, "Returned url and requested url should be the same");
    } connectTimeout:30000];

    [USRVWebRequestQueue requestUrl:targetUrl type:@"GET" headers:NULL body:NULL completeBlock:^(NSString *url, NSError *error, NSString *response, long responseCode, NSDictionary<NSString *,NSString *> *headers) {
        [completions addObject:[NSNumber numberWithInt:4]];
        XCTAssertEqual(targetUrl, url, "Returned url and requested url should be the same");
    } connectTimeout:30000];

    [USRVWebRequestQueue requestUrl:targetUrl type:@"GET" headers:NULL body:NULL completeBlock:^(NSString *url, NSError *error, NSString *response, long responseCode, NSDictionary<NSString *,NSString *> *headers) {
        [completions addObject:[NSNumber numberWithInt:5]];
        XCTAssertEqual(targetUrl, url, "Returned url and requested url should be the same");
        [expectation fulfill];
    } connectTimeout:30000];
    
    __block BOOL success = true;
    [self waitForExpectationsWithTimeout:30 handler:^(NSError * _Nullable error) {
        if (error) {
            success = false;
            XCTAssertTrue(success, "Did not complete");
        }
    }];
    
    NSNumber *previousValue = [NSNumber numberWithInt:0];
    for (NSNumber *value in completions) {
        XCTAssertTrue(previousValue < value, "Previous completion value should always be smaller than the next one");
    }
}

@end

File: UnityAdsTests/WebRequestQueueTests.swift
import XCTest

class WebRequestQueueTest: XCTestCase {
    override func setUp() {
        super.setUp()
        UADSWebRequestQueue.start()
    }
    
    func testBasicGetRequest () {
        let url = "http://www.google.fi"
        let expectation = self.expectationWithDescription("requestEndExpectation")

        UADSWebRequestQueue.requestUrl(url, type: "GET", headers: nil, body: nil, completeBlock: {
            completeBlock in
            XCTAssertEqual(completeBlock.0, url, "Returned url and requested url should be the same")
            expectation.fulfill()
            }, connectTimeout: 30000);

        self.waitForExpectationsWithTimeout(30) {
            error in
            XCTAssertTrue(true, "Didn't timeout")
        }
    }
    
    func testMultipleGetRequests () {
        let url = "http://www.google.fi"
        
        let expectation = self.expectationWithDescription("requestEndExpectation")
        var completions = [Int]()
        
        UADSWebRequestQueue.requestUrl(url, type: "GET", headers: nil, body: nil, completeBlock: {
            completeBlock in
            completions.append(1)
            XCTAssertEqual(completeBlock.0, url, "Returned url and requested url should be the same")
            }, connectTimeout: 30000);
        
        UADSWebRequestQueue.requestUrl(url, type: "GET", headers: nil, body: nil, completeBlock: {
            completeBlock in
            completions.append(2)
            XCTAssertEqual(completeBlock.0, url, "Returned url and requested url should be the same")
        }, connectTimeout: 30000)

        UADSWebRequestQueue.requestUrl(url, type: "GET", headers: nil, body: nil, completeBlock: {
            completeBlock in
            completions.append(3)
            XCTAssertEqual(completeBlock.0, url, "Returned url and requested url should be the same")
        }, connectTimeout: 30000)

        UADSWebRequestQueue.requestUrl(url, type: "GET", headers: nil, body: nil, completeBlock: {
            completeBlock in
            completions.append(4)
            XCTAssertEqual(completeBlock.0, url, "Returned url and requested url should be the same")
        }, connectTimeout: 30000)

        UADSWebRequestQueue.requestUrl(url, type: "GET", headers: nil, body: nil, completeBlock: {
            completeBlock in
            completions.append(5)
            XCTAssertEqual(completeBlock.0, url, "Returned url and requested url should be the same")
            expectation.fulfill()
        }, connectTimeout: 30000)
        
        self.waitForExpectationsWithTimeout(30) {
            error in
            XCTAssertTrue(true, "Didn't timeout")
        }
        
        var previousValue = 0
        for value in completions {
            XCTAssertTrue(previousValue < value, "Previous completion value should always be smaller than the next one")
            previousValue = value
        }
    }
}
